### DISCUSSION TITLE
How to distinguish consumers from each other during MessageReceived?

### DISCUSSION THREAD

**Discussion by plewam** [rabbitmq/rabbitmq-stream-dotnet-client#321] [General]
Hi all,

I am planning to have multiple consumers being bound to the stream system. They might consume from different streams or the same stream. How can I tell which consumer received the message during the `MessageReceived` method? I might be missing something, but when I create a consumer I can pass in a a `Reference` into the `ConsumerConfig`. Can't this `Reference` be exposed during the `MessageReceived` method?

Kind regards,

Mathis

**Comment by Gsantomaggio** [comment]
We can expose the current consumer configuration and I think that makes sense, but when you create a consumer you have different handlers like:
```csharp
            var consumer1 = await Consumer.Create(
                new ConsumerConfig(system, stream)
                {
                    Reference = "my_consumer_1",
                    MessageHandler = async (sourceStream, consumer, ctx, message) =>
                    {
                      /// handler for my_consumer_1
                    }
              ...}

   .......


            var consumer2 = await Consumer.Create(
                new ConsumerConfig(system, stream)
                {
                    Reference = "my_consumer_2",
                    MessageHandler = async (sourceStream, consumer, ctx, message) =>
                    {
                      /// handler for my_consumer_2
                    }





                        
```

**Reply by plewam** [reply]
Hi @Gsantomaggio, 
Thanks for your reply. It would be great if you could do this. I get your above idea, however during compile time I don't know yet how many consumers/ MessageHandler I am going to create. Therefore I need something to differentiate the consumer from each other during message receive.

**Reply by Gsantomaggio** [reply]
@plewam makes sense! We will expose all or part of the configuration.

**Reply by Gsantomaggio** [reply]
@plewam: https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/322 WDYT?

**Reply by plewam** [reply]
@Gsantomaggio Looks superb to me! Its ok for the that the info for the producer is deprecated in the future. The consumer is the important part.

**Reply by Gsantomaggio** [reply]
@plewam FYI https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/releases/tag/v1.7.3

**Reply by plewam** [reply]
@Gsantomaggio Perfect, thank you. I have pulled the latest version and will now implement my logic on top of it. Thanks again for your quick help :)

**Comment by Gsantomaggio** [comment]
@plewam this  https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/348 adds more info. I think can be interesting for you.
